src/Datadog.Trace/TracingProcessManager.cs

```diff
@@ -115,65 +123,112 @@ private static Task StartProcessWithKeepAlive(ProcessMetadata metadata)
                                 return;
                             }
 
-                            try
+                            if (metadata.SequentialFailures >= MaxFailures)
                             {
-                                metadata.IsFaulted = false;
+                                Log.Error("Circuit breaker triggered for {Process}. Max failed retries reached ({ErrorCount}).", path, MaxFailures);
+                                metadata.ProcessState = ProcessState.Faulted;
+                                return;
+                            }
 
-                                if (metadata.ProgramIsRunning())
+                            try
+                            {
+                                if (metadata.ProcessState == ProcessState.NotRunning)
                                 {
-                                    Log.Debug("{Process} is already running.", path);
-                                    continue;
+                                    // This means we have never tried from this domain
+                                    var pipeIsBound = metadata.NamedPipeIsBound();
+
+                                    if (pipeIsBound)
+                                    {
+                                        // It is possible that if a pipe is bound it may still be shutting down from last time
+                                        // Check on a delay to be sure we have the agent available
+                                        var attempts = 7;
+                                        var delay = 50d;
+
+                                        while (--attempts > 0)
+                                        {
+                                            await Task.Delay((int)delay);
+                                            if (!metadata.NamedPipeIsBound())
+                                            {
+                                                metadata.ProcessState = ProcessState.ReadyToStart;
+                                                break;
+                                            }
+
+                                            // Should result in a max delay of ~3.28 seconds before giving up
+                                            delay = delay * 1.75d;
+                                        }
+
+                                        metadata.ProcessState = ProcessState.Healthy;
+                                    }
+                                    else
+                                    {
+                                        // If no pipe is bound, kick it off
+                                        metadata.ProcessState = ProcessState.ReadyToStart;
+                                    }
```

Comment:
So if the pipe _is_ bound, and remains so, does that mean we'll lose 3+ seconds at the start? That seems like quite a lot? Or do we use the pipe if it's available?

Reply:
We use the pipe if it's available.
Currently, if there is a conflict on the first try, we lose stats for ~10 seconds.
This conflict arises if a named pipe has not been cleaned up from a previous instance.
The first process start fails because the named pipe is not available to bind to, and the next retry doesn't come through for 10 seconds, even though the pipe might be cleared in 100ms.

This mechanism is to check the named pipe in quicker intervals on startup.
If a named pipe is bound, it will check in 50ms, then 87ms, then 153 ms, and so on until a max wait of 1436ms, which sums to approximately 3.28 seconds.

Another improvement here which is less obvious, the first failed attempt goes away.
Previously, we were deciding to start if the process was not running, but the named pipe remains around after process kill in testing, so the named pipe being bound is a more accurate timing check.
This means no more (or at least incredibly rare) transient error messages on startup.

Reply:
Ah ok, I think the point I was missing is that there's (presumably) no way for a pipe to transition from bound -> unbound -> bound inside the `ProcessState.NotRunning` branch. You'll only ever see a bound -> unbound transition (as it's rebound in the `ProcessState.Healthy || ProcessState.Faulted` branch).

Reply:
Unless there are multiple processes or even app domains running. :)

Reply:
ok... then if that _is_ a possibility, (bound -> unbound -> bound), then there's potential for a race condition here where you wait for 3s unnecessarily right? I'm not sure if a) if that's even avoidable without the long lived process approach, b) if it's a problem 🙂

Reply:
Yes, any separate process or app domain will run into this delay for this task.
This is totally acceptable to me, as it doesn't block the application and makes the agent processes much more robust.